packages/block-editor/src/components/list-view/use-list-view-drop-zone.js

```diff
@@ -52,19 +58,91 @@ import { store as blockEditorStore } from '../../store';
  *                                                  'inside' refers to nesting as an inner block.
  */
 
+export const NESTING_LEVEL_INDENTATION = 28;
+
+/**
+ * Determines whether the user is positioning the dragged block to be
+ * moved up to a parent level.
+ *
+ * Determined based on nesting level indentation of the current block.
+ *
+ * @param {WPPoint} point        The point representing the cursor position when dragging.
+ * @param {DOMRect} rect         The rectangle.
+ * @param {number}  nestingLevel The nesting level of the block.
+ * @return {boolean} Whether the gesture is an upward gesture.
+ */
+function isUpGesture( point, rect, nestingLevel = 1 ) {
+	// If the block is nested, and the user is dragging to the bottom
+	// left of the block, then it is an upward gesture.
+	const blockIndentPosition =
+		rect.left + nestingLevel * NESTING_LEVEL_INDENTATION;
+	return point.x < blockIndentPosition;
+}
+
+function getDesiredRelativeParentLevel( point, rect, nestingLevel = 1 ) {
+	const blockIndentPosition =
+		rect.left + nestingLevel * NESTING_LEVEL_INDENTATION;
+	const desiredParentLevel = Math.round(
+		( point.x - blockIndentPosition ) / NESTING_LEVEL_INDENTATION
+	);
+	return Math.abs( desiredParentLevel );
+}
+
+/**
+ * Returns an array of the parent blocks of the block the user is dropping to.
+ *
+ * @param {WPListViewDropZoneBlock}  candidateBlockData The block the user is dropping to.
+ * @param {WPListViewDropZoneBlocks} blocksData         Data about the blocks in list view.
+ * @return {WPListViewDropZoneBlocks} An array of block parents, including the block the user is dropping to.
+ */
+function getCandidateBlockParents( candidateBlockData, blocksData ) {
+	const candidateBlockParents = [];
+	let currentBlockData = candidateBlockData;
+
+	while ( currentBlockData ) {
+		candidateBlockParents.push( { ...currentBlockData } );
+		currentBlockData = blocksData.find(
```

Comment:
Is it worth updating `getListViewDropTarget` and all its helper functions to have `blocksData` be a `Map` (keyed by `clientId`) instead of an array so that we don't need lookups like this `find()`? I'm not too familiar with this code but it looks like `getCandidateBlockParents` will be called when the user moves the mouse so I imagine performance could be a problem if there are ~1000s of blocks.

Only bother if it makes an actual difference 😀 (premature optimisation yada yada)

Reply:
When I originally wrote it, I'm sure `blockData` was supposed to only be computed once when the user starts dragging. I see it's now computed on every dragover event. 🤔 

Anyway, my suggestion was going to be to store the parent client ids in `blocksData`, but I think the only real optimization would come from ensuring that `blocksData` isn't computed so often.

Reply:
I tried a fix for this in this PR - https://github.com/WordPress/gutenberg/pull/50210

Reply:
Thanks Dan! I like that direction — just left a comment on that PR about whether we then need to add data attributes to the placeholder rows that are used as part of the windowing logic. I think that should still be pretty doable, while preserving the performance improvement?